tests/unit/test_charm.py

```diff
@@ -1,105 +1,138 @@
-    setup_grpc_relation(harness, "grpc-two", "9090")
-    # In order to avoid the charm going to Blocked
-    setup_ingress_relation(harness)
-    harness.begin_with_initial_hooks()
+    def test_with_grpc_relation(self, harness):
+        """Test that the grpc Component is active when one grpc relation is present."""
+        harness.set_leader(True)
+        setup_grpc_relation(harness, "grpc-one", "8080")
+        harness.begin_with_initial_hooks()
 
-    pod_spec, _ = harness.get_pod_spec()
+        assert isinstance(harness.charm.grpc.status, ActiveStatus)
 
-    expected = yaml.safe_load(open("tests/unit/many_relations.yaml"))
+    def test_with_ingress_relation(self, harness):
+        """Test that the ingress_relation Component is active when an ingress is present."""
+        harness.set_leader(True)
+        # Set required grpc relation
+        setup_grpc_relation(harness, "grpc-one", "8080")
+        setup_ingress_relation(harness)
 
-    c = pod_spec["containers"][0]["volumeConfig"][0]["files"][0]["content"]
-    assert json.loads(c) == expected
+        harness.begin_with_initial_hooks()
 
-    assert harness.charm.model.unit.status == ActiveStatus("")
+        assert harness.charm.ingress_relation.status == ActiveStatus()
 
+    def test_envoy_config_generator(self, harness):
+        """Test the envoy_config_generator Component is active when prerequisites are ready."""
+        harness.set_leader(True)
+        setup_grpc_relation(harness, "grpc-one", "8080")
+        setup_ingress_relation(harness)
 
-def test_with_ingress_relation(harness):
-    harness.set_leader(True)
-    add_oci_image(harness)
+        harness.begin_with_initial_hooks()
 
-    # Set required grpc relation
-    setup_grpc_relation(harness, "grpc-one", "8080")
+        assert harness.charm.envoy_config_generator.status == ActiveStatus()
 
-    setup_ingress_relation(harness)
+    def test_envoy_config_generator_if_grpc_get_data_returns_empty_dict(self, harness):
+        """Test the envoy_config_generator Component raises an error if grpc data is not ready."""
+        harness.set_leader(True)
+        setup_grpc_relation(harness, "grpc-one", "8080")
+        setup_ingress_relation(harness)
 
-    harness.begin_with_initial_hooks()
+        # Let the charm start up with the grpc relation
+        harness.begin_with_initial_hooks()
 
-    assert isinstance(harness.charm.model.unit.status, ActiveStatus)
+        # Now mock out the grpc relation's get_data to return an empty dict
+        harness.charm.grpc.component.get_data = lambda: {}
 
+        assert isinstance(harness.charm.envoy_config_generator.status, BlockedStatus)
 
-# Helper functions
-def add_oci_image(harness: Harness):
-    harness.add_oci_resource(
-        "oci-image",
-        {
-            "registrypath": "ci-test",
-            "username": "",
-            "password": "",
-        },
-    )
+        # Do something to trigger a config changed to update the charm status
```

Comment:
I think there's two threads to this discussion:

### Blocked vs Error

I see where you're coming from, but the `CharmReconciler` is going to prevent the charm from going into `Error` in either case.  

`CharmReconciler` catches errors thrown by any `Component` and handles them by logging and setting their `Component` to `Blocked` (so in this case, something like `BlockedStatus('[envoy_config_generator] SOMETHING')`. The motivation behind this is to prevent one erroring Component of a charm from taking down the entire charm if you don't want it to (the case in mind at the time was istio, where having an error when setting up an ingress shouldn't prevent someone from being able to enable auth on the gateway, etc)

If we want these to really throw errors to Juju, we'd have to change something in `CharmReconciler`  I think.  

### the assertion after the `config-changed`

the assertion after the unrelated `config-changed` is that the unit is **not** active.  Basically just confirming that, now that `get_data` is broken, the charm is not showing as active

Reply:
Re blocked vs error:
In this case, though, is there anything that a user can do to unblock the charm? That's why to me it is more appropriate to error out as this is probably an issue with the charm or the related applications. If we want to avoid erroring out, because of the CharmReconciler limitations, we could put the unit in maintenance or event waiting status, but log all of the details as this error seems unrecoverable.

> the assertion after the unrelated config-changed is that the unit is not active. Basically just confirming that, now that get_data is broken, the charm is not showing as active

Got it. The comment message was a bit misleading, it says "Do something to trigger a config changed to update the charm status", I thought the unit status would've changed on a config changed event. Anyway, LGTM.

Reply:
Thinking about this a bit more, I see the limitations that the `CharmReconciler` has, we probably want to discuss about it, no reason to block this PR. Thanks for the clarification @ca-scribner !